packages/ceramic-core/src/document.ts

```diff
@@ -179,38 +202,48 @@ class Document extends EventEmitter {
       const cid = entry.value[1]
       const record = await this.dispatcher.retrieveRecord(cid)
       // TODO - should catch potential thrown error here
-      state = await this._doctype.applyRecord(record, cid, state)
+      if (!record.next) {
+        state = await this._doctypeHandler.applyGenesis(record, cid)
+      } else if (record.proof) {
+        const proof = await this._verifyAnchorRecord(record)
+        state = await this._doctypeHandler.applyAnchor(record, proof, cid, state)
+      } else {
+        state = await this._doctypeHandler.applySigned(record, cid, state)
+      }
       if (breakOnAnchor && state.anchored) return state
       entry = itr.next()
     }
     return state
   }
 
+  async _verifyAnchorRecord (record: AnchorRecord): Promise<AnchorProof> {
+    //const nextRecordA = await this.dispatcher.retrieveRecord(record.next)
+    //const nextRecordB = await this.dispatcher.retrieveRecord(record.proof + '/root' + record.path)
+    // assert A == B
+    const proof: AnchorProof = await this.dispatcher.retrieveRecord(record.proof)
+    await this._anchorService.validateChainInclusion(proof)
+    return proof
+  }
+
   async _publishHead (): Promise<void> {
     await this.dispatcher.publishHead(this.id, this.head)
   }
 
   async change (newContent: any): Promise<boolean> {
-    const record = await this._doctype.makeRecord(this._state, newContent)
+    const record = await this._doctypeHandler.makeRecord(this._state, newContent)
     const cid = (await this.dispatcher.storeRecord(record)).toString()
-    this._state = await this._doctype.applyRecord(record, cid, this._state)
+    this._state = await this._doctypeHandler.applySigned(record, cid, this._state)
     await this.anchor()
     this._publishHead()
     return true
   }
 
-  async anchor (): Promise<boolean> {
-    // fake anchor
-    const proof = {
-      chain: 'ethmainnet',
-      blockNumber: Date.now(),
-      txHash: 'eth-cid',
-      root: 'cid'
-    }
-    const record = { proof, path: 'ipld path for witness', next: { '/': this.head } }
-    const cid = (await this.dispatcher.storeRecord(record)).toString()
-    this._state = await this._doctype.applyRecord(record, cid, this._state)
-    return true
+  async anchor (): Promise<void> {
+    this._anchorService.on(this.id, async (cid): Promise<void> => {
```

Comment:
Where is this listener cleaned up?

Reply:
It's removed on line 28 of `packages/ceramic-core/src/anchor-service.ts`.